aopy/data/bmi3d.py

```diff
@@ -1117,6 +1117,91 @@ def get_ts_data_segment(preproc_dir, subject, te_id, date, start_time, end_time,
+        else:
+            spike_segment[unit_label], bins = precondition.bin_spike_times(temp_spike_segment, start_time, end_time, bin_width)
+
+    return spike_segment, bins
+
+def get_spike_data_aligned(preproc_dir, subject, te_id, date, port, trigger_times, time_before, time_after, bin_width=0.01):
+    """
+    Loads spike data for a given subject and experiment, then aligns binned spike to trigger times.
+
+    .. image:: _images/spike_align_example.png
+
+    Args:
+        preproc_dir (str): Path to the preprocessed data directory.
+        subject (str): Subject name.
+        te_id (str): Task entry number.
+        date (str): The date of the experiment.
+        port (int): The port number corresponding to the spike data.
+        trigger_times (numpy.ndarray): 1D Array of trigger times (in seconds) for each trial to which spike data should be aligned.
+        time_before (float): The amount of time (in seconds) before each trigger time to include in the aligned spike data.
+        time_after (float): The amount of time (in seconds) after each trigger time to include in the aligned spike data.
+        bin_width (float, optional): The width of the bins [s]. Default is 0.01 (10ms) seconds. 
+
+    Returns:
+        tuple: A tuple containing:
+            - spike_aligned (numpy.ndarray): A 3D array of aligned spike data with shape (ntime, nunits, ntrials), where:
+                - ntime is the number of time bins between `time_before` and `time_after` around each trigger.
+                - nch is the number of units.
+                - ntrials is the number of trials (trigger events).
+            - unit_labels (list of str): A list of unit labels corresponding to the 'nunits' dimension in the aligned spike data.
+            - bins (numpy.ndarray): The time bin centers relative to the trigger times.
+    """
+    # Load data
+    filename_mc = get_preprocessed_filename(subject, te_id, date, 'spike')
+    spike_data = load_hdf_group(os.path.join(preproc_dir, subject), filename_mc, f'drive{port}/spikes')
+    
+    # Define relevant variables
+    samplerate = int(np.round(1/bin_width))
+    bins = np.arange(-time_before, time_after, bin_width) + bin_width/2
+    ntime = int(np.round((time_after+time_before)/bin_width))
+    nch = len(spike_data)
+    ntrials = len(trigger_times)
+    
+    # Parse segment and bin spikes if necessary.
+    unit_labels = list(spike_data.keys())
+    spike_aligned = np.zeros((ntime, nch, ntrials))*np.nan #(ntime, nch, ntrials)
+    for iunit, unit_label in enumerate(unit_labels):
+        binned_spikes, _ = precondition.bin_spike_times(spike_data[unit_label], 0, trigger_times[-1]+time_after, bin_width)
+        spike_aligned[:,iunit,:] = np.squeeze(preproc.base.trial_align_data(binned_spikes, trigger_times-(bin_width/2), time_before, time_after, samplerate)) # Squeeze to remove singleton dimension from only doing one unit at a time. Adjust trigger times to center on the previously binned spikes. 
```

Comment:
can you explain why you chose to do binning and then aligning rather than align first and then bin?

Reply:
also what's the utility of this function that operates on a single file rather than a version that can take multiple recordings as input similar to tabulate_spike_data_segments below?

Reply:
I copied the same structure that I saw used for lfp data. "get_lfp_aligned" also only operates on one function. I could make a wrapper for this but I think there are a few things that are unclear now:

- what output is best. A concatenated array? A list of arrays? etc. 
- How would metadata be tracked (unit labels, etc)? Would we need a whole dictionary?

For these reasons, I've decided to wait until more people are using the data to see what the best path forward is.

Reply:
ok makes sense!